test/lit/gc-read-write-effects.wast

```diff
@@ -0,0 +1,73 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
+
+;; Check that writing a struct field is not reordered with reading the same
+;; struct field.
+
+;; RUN: wasm-opt -all -O1 %s -S -o - | filecheck %s
+
+(module
+ (type $A (struct
+   (field $i (mut i32))
+   (field $b (mut (ref null $B)))
+ ))
+ (type $B (struct
+   (field $i (mut i32))
+   (field $c (mut (ref null $C)))
+   (field $b (mut (ref null $B)))
+ ))
+ (type $C (struct
+   (field (mut i32))
+   (field (mut i64))
+ ))
+ (export "f" (func $f))
+
+ ;; Check that this:
+ ;;
+ ;;   c = a.b.c
+ ;;   a.b = a.b.b
+ ;;   return c
+ ;;
+ ;; Is not turned into this:
+ ;;
+ ;;   a.b = a.b.b
+ ;;   return a.b.c
+ ;;
+ ;; CHECK:      (func $f (param $0 (ref null $A)) (result (ref null $C))
+ ;; CHECK-NEXT:  (local $1 (ref null $C))
+ ;; CHECK-NEXT:  (local.set $1
+ ;; CHECK-NEXT:   (struct.get $B $c
+ ;; CHECK-NEXT:    (struct.get $A $b
+ ;; CHECK-NEXT:     (local.get $0)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (struct.set $A $b
+ ;; CHECK-NEXT:   (local.get $0)
+ ;; CHECK-NEXT:   (struct.get $B $b
+ ;; CHECK-NEXT:    (struct.get $A $b
+ ;; CHECK-NEXT:     (local.get $0)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT:  (local.get $1)
+ ;; CHECK-NEXT: )
```

Comment:
side note, how do you write such an output? by hand, adding `;; CHECK-NEXT` etc., or with a tool?

Reply:
For this file I just ran scripts/update_lit_checks.py! For other files where I am checking something besides function contents, I write them by hand.

Reply:
Where "by hand" means liberal copy-paste and sometime emacs macros.

Reply:
Got it, thanks!